### PR TITLE
[adhoc-filters] Adding adhoc-filters to all viz types

### LINKED ISSUE #5172: Filters not saved when visualization changed
Make sure these boxes are checked before submitting your issue - thank you!

- [x ] I have checked the superset logs for python stacktraces and included it here as text if any
- [ x] I have reproduced the issue with at least the latest released version of superset
- [x ] I have checked the issue tracker for the same issue and I haven't found one similar

### Superset version
master

### Expected results
When exploring a chart, filters are saved when changing visualizations.

### Actual results
If you are exploring a chart with the Time Series Line Chart visualization and have filters defined, they are removed if you switch to Table View. (And, the filter style reverts to the old version). This is a regression -- the filters used to be saved.

### Steps to reproduce
Create a chart with a Time Series Line Char viz.
Add filters.
Change the chart to Table View.
Note that the filters disappear.

### PR DESCRIPTION
**Pull request by john-bodley**
So this PR came about because I was working on another feature which required a database migration to update SQL filters when I realized that I needed to update multiple form-data fields, specifically the following fields exist around filters:

- adhoc_filters
- filters
- having
- having_filters
- where
 
Hence the purpose of this PR is to condense the number of form-data fields and simplify filter logic by rolling out ad-hoc filters for all visualization types. 

Note I sense it's worthwhile to first outline how the current situation works. On the front-end the ad-hoc filters take the legacy `filters`, `having`, `having_filters`, and `where` fields which are then merged with `adhoc_filters`. The `adhoc_filters` persist in the URL form-data (and when the slice is saved) though are transformed (and removed) back into the legacy filters in the backend. This means that all five filters types can actually persist in the form-data.

This PR is the first of a two-phase change (to help minimize code change). This change is mostly a front-end change, i.e., it applies ad-hoc filters to all visualization types and performs a database migration to upgrade the filters (as opposed to doing this in the frontend). Note the backend still transforms the ad-hoc filters to legacy filters (though the `adhhoc_filters` field persists). Although this represents a duplication of data the ad-hoc filters are the source of truth throughout, i.e., the legacy filters are always derived from the ad-hoc filters. Note this change ensures that old URLs with legacy filters will still work.

For phase II the plan is to clean up the backend and deprecate the legacy filters. It will ensure URLs with legacy filters will continue to work.   

Finally the ad-hoc filters include a unique ID which is required for the frontend which gets persisted to the database. There may be merit in removing this in the future.

Closes https://github.com/apache/incubator-superset/issues/5172.

to: @betodealmeida @GabeLoins @michellethomas @mistercrunch @williaster

### CODE REVIEW

File: superset/utils.py [hunk 1]
```diff
@@ -877,8 +877,15 @@ def split_adhoc_filters_into_base_filters(fd):
                     sql_where_filters.append(adhoc_filter.get('sqlExpression'))
                 elif clause == 'HAVING':
                     sql_having_filters.append(adhoc_filter.get('sqlExpression'))
-        fd['where'] = ' AND '.join(['({})'.format(sql) for sql in sql_where_filters])
-        fd['having'] = ' AND '.join(['({})'.format(sql) for sql in sql_having_filters])
-        fd['having_filters'] = simple_having_filters
-        fd['filters'] = simple_where_filters
-        del fd['adhoc_filters']
```
Comment (john-bodley) [inline superset/utils.py]: Note we can no longer remove the `adhoc_filters` field as we need to ensure that ad-hoc filters persist. Note currently by deleting this field the ad-hoc filters actually get rebuilt on the frontend by merging the legacy and ad-hoc filters.

File: tests/viz_tests.py [hunk 2]
```diff
@@ -259,7 +259,7 @@ def test_adhoc_filters_overwrite_legacy_filters(self):
             query_obj['extras']['having_druid'],
         )
         self.assertEqual('(value3 in (\'North America\'))', query_obj['extras']['where'])
-        self.assertEqual('', query_obj['extras']['having'])
+        self.assertEqual('having': 'SUM(value1) > 5', query_obj['extras']['having'])
```
Comment (john-bodley) [inline tests/viz_tests.py]: @GabeLoins I'm unsure why `having` was supposed to be an empty string previously.

File: superset/utils.py [hunk 3]
```diff
@@ -881,4 +881,3 @@ def split_adhoc_filters_into_base_filters(fd):
         fd['having'] = ' AND '.join(['({})'.format(sql) for sql in sql_having_filters])
         fd['having_filters'] = simple_having_filters
         fd['filters'] = simple_where_filters
-        del fd['adhoc_filters']
```
Comment (john-bodley) [inline superset/utils.py]: Note we can no longer remove the `adhoc_filters` field as we need to ensure that ad-hoc filters persist. Note currently by deleting this field the ad-hoc filters actually get rebuilt on the frontend by merging the legacy and ad-hoc filters.

File: tests/viz_tests.py [hunk 4]
```diff
@@ -259,7 +259,7 @@ def test_adhoc_filters_overwrite_legacy_filters(self):
             query_obj['extras']['having_druid'],
         )
         self.assertEqual('(value3 in (\'North America\'))', query_obj['extras']['where'])
-        self.assertEqual('', query_obj['extras']['having'])
+        self.assertEqual('SUM(value1) > 5', query_obj['extras']['having'])
```
Comment (john-bodley) [inline tests/viz_tests.py]: @GabeLoins I'm unsure why having was supposed to be an empty string previously.

File: superset/assets/backendSync.json [hunk 5]
```diff
@@ -1683,18 +1683,6 @@
       "renderTrigger": true,
       "default": ""
     },
-    "where": {
```
Comment (GabeLoins) [inline superset/assets/backendSync.json]: I haven't touched this file before- what is it doing exactly?

File: superset/assets/spec/javascripts/explore/components/AdhocFilterControl_spec.jsx [hunk 6]
```diff
@@ -68,49 +59,6 @@ describe('AdhocFilterControl', () => {
     expect(wrapper.find(OnPasteSelect)).to.have.lengthOf(1);
   });
 
-  it('will translate legacy filters into adhoc filters if no adhoc filters are present', () => {
```
Comment (GabeLoins) [inline superset/assets/spec/javascripts/explore/components/AdhocFilterControl_spec.jsx]: bye specs!

File: superset/assets/src/components/AlteredSliceTag.jsx [hunk 7]
```diff
@@ -56,13 +60,15 @@ export default class AlteredSliceTag extends React.Component {
       return 'N/A';
     } else if (value === null) {
       return 'null';
-    } else if (controls[key] && controls[key].type === 'FilterControl') {
+    } else if (controls[key] && controls[key].type === 'AdhocFilterControl') {
```
Comment (GabeLoins) [inline superset/assets/src/components/AlteredSliceTag.jsx]: was this line necessary before? Should I have already made this change or was it only necessary once we removed the legacy filters?

**Review by gabe-lyons — APPROVED** [review]
Looks good from my end! Thank you john for making this improvement!

File: superset/assets/src/components/AlteredSliceTag.jsx (hunk 7 above)
Comment (john-bodley) [reply superset/assets/src/components/AlteredSliceTag.jsx]: @GabeLoins I'm not certain whether ad-hoc filters actually showed up in the altered slice view.

File: superset/assets/backendSync.json (hunk 5 above)
Comment (mistercrunch) [reply superset/assets/backendSync.json]: That represents the old TextControl where people would write their WHERE clause. Did we even run a db migration to convert the old filters (where & having) to new ones?

Comment (mistercrunch) [reply superset/assets/backendSync.json]: Oh I see it's in this PR !

Comment (john-bodley) [reply superset/assets/backendSync.json]: @mistercrunch I'm hoping to clean up the backend later this week so we can remove all references to the old where, having clauses entirely.

**Pull request merged**